### PR TITLE
Add issues to the Finish RAMP Intake screen

### PR DESCRIPTION
**Pull request by shanear**
The formatting is a little off, because VACOLS issues are complicated and terrible. But this gets everything on the screen:

screen shot 2017-10-19 at 11 17 10 am

Additionally, we are keeping the info icon in because that adheres to the styleguide.

connects #3583

### CODE REVIEW

File: app/models/appeal.rb [hunk 1]
```diff
@@ -3,7 +3,6 @@ class Appeal < ActiveRecord::Base
   include AssociatedVacolsModel
   include CachedAttributes
   include RegionalOfficeConcern
-  include CachedAttributes
```
Comment (shanear) [inline app/models/appeal.rb]: sad to see this go. Hope everything still works

File: app/models/concerns/cached_attributes.rb [hunk 2]
```diff
@@ -16,16 +16,16 @@ def get_cached_value(attr_name)
...
+    Rails.cache.write(cache_id(attr_name), value, write_options) && value
   end
 
   def cache_id(attr_name)
     "#{self.class.name}-#{id}-cached-#{attr_name}"
   end
 
   module ClassMethods
-    def cache_attribute(attr_name, &get_value)
+    def cache_attribute(attr_name, write_options = {}, &get_value)
```
Comment (shanear) [inline app/models/concerns/cached_attributes.rb]: this allows us to specify the cache expiration time on a per attribute basis

File: app/models/ramp_intake.rb [hunk 3]
```diff
@@ -20,6 +22,16 @@ def cancel!
     end
   end
 
+  cache_attribute :cached_serialized_appeal_issues, expires_in: 10.minutes do
```
Comment (shanear) [inline app/models/ramp_intake.rb]: 10 minutes was slightly arbitrary, but probably safe

File: client/app/intake/redux/reducer.js [hunk 4]
```diff
@@ -3,6 +3,20 @@ import { update } from '../../util/ReducerUtil';
 import { formatDateStr } from '../../util/DateUtil';
 import _ from 'lodash';
 
+const formatAppeals = (appeals) => {
+  return _.map(appeals, (appeal) => (
+    {
+      issues: _.map(appeal.issues,
```
Comment (shanear) [inline client/app/intake/redux/reducer.js]: not sure if we're using `_.map` or `Array.map`.

File: lib/fakes/appeal_repository.rb [hunk 5]
```diff
@@ -170,6 +170,7 @@ def self.seed!(app_name: nil)
     seed_certification_data! if app_name.nil? || app_name == "certification"
     seed_establish_claim_data! if app_name.nil? || app_name == "dispatch-arc"
     seed_reader_data! if app_name.nil? || app_name == "reader"
+    seed_intake_data! if app_name.nil? || app_name == "intake"
```
Comment (shanear) [inline lib/fakes/appeal_repository.rb]: This is the proper place to put Fakes seed data.

File: lib/fakes/initializer.rb [hunk 6]
```diff
@@ -51,16 +51,6 @@ def load_fakes_and_seed!(app_name: nil)
 
       Fakes::AppealRepository.seed!(app_name: app_name)
       Fakes::HearingRepository.seed! if app_name.nil? || app_name == "hearings"
-
-      load_ramp_election_fakes
```
Comment (shanear) [inline lib/fakes/initializer.rb]: this is not the right place for Fakes seed data

File: lib/generators/appeal.rb [hunk 7]
```diff
@@ -170,6 +170,7 @@ def build(attrs = {})
       vacols_record = extract_vacols_record(attrs)
       appeal = Appeal.find_or_initialize_by(vacols_id: attrs[:vacols_id])
       inaccessible = attrs.delete(:inaccessible)
+      veteran = attrs.delete(:veteran)
```
Comment (shanear) [inline lib/generators/appeal.rb]: this allows the test to specify it's own veteran, instead of auto generating one

**Comment by NickHeiner** [comment]
I get a 404 on this branch.

image

I don't get it on `master`.

File: client/app/intake/pages/finish.jsx [hunk 8]
```diff
@@ -3,18 +3,46 @@ import Button from '../../components/Button';
...
 import { completeIntake, confirmFinishIntake } from '../redux/actions';
 import { bindActionCreators } from 'redux';
 import { getRampElectionStatus } from '../redux/selectors';
+import _ from 'lodash';
 
 const submitText = "I've completed all steps";
 
 class Finish extends React.PureComponent {
+  issuesAlertContent = (appeals) => {
+    let issueColumns = [
```
Comment (NickHeiner) [inline client/app/intake/pages/finish.jsx]: Please default to `const`.

File: client/app/intake/pages/finish.jsx [hunk 9]
```diff
@@ -3,18 +3,46 @@ import Button from '../../components/Button';
...
 import { connect } from 'react-redux';
 import { completeIntake, confirmFinishIntake } from '../redux/actions';
 import { bindActionCreators } from 'redux';
 import { getRampElectionStatus } from '../redux/selectors';
+import _ from 'lodash';
 
 const submitText = "I've completed all steps";
 
 class Finish extends React.PureComponent {
+  issuesAlertContent = (appeals) => {
```
Comment (NickHeiner) [inline client/app/intake/pages/finish.jsx]: This method should be named with a verb, like `getIssuesAlertContent`.

File: client/app/intake/pages/finish.jsx [hunk 10]
```diff
@@ -3,18 +3,46 @@ import Button from '../../components/Button';
...
+        )
+      },
+      {
+        header: 'Note',
+        valueName: 'note'
+      }
+    ];
+
+    return _.map(appeals, (appeal) => {
+      return <div>
```
Comment (NickHeiner) [inline client/app/intake/pages/finish.jsx]: Why do we have this div?

File: client/app/intake/pages/finish.jsx [hunk 11]
```diff
@@ -50,10 +78,18 @@ class Finish extends React.PureComponent {
       () => <span><strong>Step {index + 1}.</strong> {step}</span>
     );
 
+    const issuesAlertTitle = `This Veteran has ${appeals.length} ` +
+                             'active appeal(s), with the following issues';
```
Comment (NickHeiner) [inline client/app/intake/pages/finish.jsx]: How about we do the pluralization correctly for `appeal(s)`? I'm happy to do that in a separate PR if you don't want to do it here.

File: client/app/intake/redux/reducer.js (hunk 4 above)
Comment (NickHeiner) [reply client/app/intake/redux/reducer.js]: We prefer the native `Array.prototype.map`, unless there's a feature in `_.map` that we need.

File: client/app/intake/pages/finish.jsx (hunk 10 above)
Comment (shanear) [reply client/app/intake/pages/finish.jsx]: for the line break. it's how dispatch did it and I felt like it was fine.

File: client/app/intake/pages/finish.jsx (hunk 8 above)
Comment (shanear) [reply client/app/intake/pages/finish.jsx]: 👍

File: client/app/intake/pages/finish.jsx (hunk 9 above)
Comment (shanear) [reply client/app/intake/pages/finish.jsx]: 👍

File: client/app/intake/pages/finish.jsx (hunk 11 above)
Comment (shanear) [reply client/app/intake/pages/finish.jsx]: ill give it a shot. I didn't know it was possible/easy.

File: client/app/intake/pages/finish.jsx (hunk 10 above)
Comment (NickHeiner) [reply client/app/intake/pages/finish.jsx]: It is not necessary. I have the following, and the table looks the same:

```jsx
    return _.map(
      appeals, 
      (appeal) => <Table columns={issueColumns} rowObjects={appeal.issues} summary="Appeal issues" />
    );
```

File: client/app/intake/pages/finish.jsx [hunk 12]
```diff
@@ -3,18 +3,46 @@ import Button from '../../components/Button';
...
+          issue.description.map((descriptor) => <div>{descriptor}</div>)
+        )
+      },
+      {
+        header: 'Note',
+        valueName: 'note'
+      }
+    ];
+
+    return _.map(appeals, (appeal) => {
```
Comment (NickHeiner) [inline client/app/intake/pages/finish.jsx]: If we are returning an array of elements, then each one needs to have a `key` set. Can we add an `id` from the backend so we can set `key={appeal.id}` here?

Comment (NickHeiner) [reply client/app/intake/pages/finish.jsx]: Here's the error that helped me find this:

image

Comment (shanear) [reply client/app/intake/pages/finish.jsx]: 👍

File: client/app/intake/pages/finish.jsx [hunk 13]
```diff
@@ -3,18 +3,46 @@ import Button from '../../components/Button';
...
+      return <div>
+        <Table columns={issueColumns} rowObjects={appeal.issues} summary="Appeal issues" />
```
Comment (NickHeiner) [inline client/app/intake/pages/finish.jsx]: This component is missing a prop:

image

File: client/app/intake/pages/finish.jsx [hunk 14]
```diff
@@ -50,10 +78,18 @@ class Finish extends React.PureComponent {
...
 
+    const issuesAlertTitle = `This Veteran has ${appeals.length} ` +
+                             'active appeal(s), with the following issues';
+
     return <div>
       <h1>Finish processing { optionName } election</h1>
       <p>Please complete the following steps outside Caseflow.</p>
       <BareOrderedList className="cf-steps-outside-of-caseflow-list" items={stepFns} />
+
+      <Alert title={ issuesAlertTitle } type="info">
```
Comment (NickHeiner) [inline client/app/intake/pages/finish.jsx]: This causes problems because `Alert` will put its content in a `<p>` tag, which does not accept arbitrary nested elements:

image
image

We should change `Alert` to accommodate this.

File: client/app/intake/redux/reducer.js (hunk 4 above)
Comment (shanear) [reply client/app/intake/redux/reducer.js]: 👍

File: client/app/intake/pages/finish.jsx (hunk 14 above)
Comment (shanear) [reply client/app/intake/pages/finish.jsx]: 👍

**Comment by NickHeiner** [comment]
I see that the model spec fails when the code is broken.

File: app/models/appeal.rb (hunk 1 above)
Comment (NickHeiner) [reply app/models/appeal.rb]: :smile:

File: client/app/intake/pages/finish.jsx (hunk 14 above)
Comment (shanear) [reply client/app/intake/pages/finish.jsx]: changed to a div. Will have no impact on styling, since the `<p>` padding and margin were all zeroed out.

Comment (shanear) [reply client/app/intake/pages/finish.jsx]: before:
screen shot 2017-10-19 at 2 12 38 pm

after:
screen shot 2017-10-19 at 2 06 19 pm

Comment (NickHeiner) [reply client/app/intake/pages/finish.jsx]: The `Alert` you posted a screenshot of is not this `Alert`. This `Alert` is the one that contains the issues table.

File: client/app/intake/redux/reducer.js [hunk 15]
```diff
@@ -3,6 +3,20 @@ import { update } from '../../util/ReducerUtil';
 import { formatDateStr } from '../../util/DateUtil';
 import _ from 'lodash';
 
+const formatAppeals = (appeals) => {
```
Comment (NickHeiner) [inline client/app/intake/redux/reducer.js]: This can be simpler:

```js
const formatAppeals = (appeals) =>
  _.map(appeals, (appeal) => {
    appeal.issues.map(
      ({ program_description, ...rest }) => ({
        programDescription: program_description,
        ...rest
      })
    );
  });
```

File: spec/models/ramp_intake_spec.rb [hunk 16]
```diff
@@ -84,6 +84,67 @@
...
+          issues: [
+            Generators::Issue.build(description: ["15 - Last Issue"], note: "")
+          ]
+        )
+      ]
+    end
+
+    it do
+      is_expected.to eq([{
+                          issues: [{
```
Comment (NickHeiner) [inline spec/models/ramp_intake_spec.rb]: Is it unusual that the issues have the same `program_description` and `description` fields? Or would we expect to see that in real data?

File: client/app/intake/pages/finish.jsx (hunk 14 above)
Comment (shanear) [reply client/app/intake/pages/finish.jsx]: i'm updating all alerts

File: client/app/components/Alert.jsx [hunk 17]
```diff
@@ -42,8 +42,8 @@ export default class Alert extends React.Component {
     return <div className={className} {...this.getRole()}>
       <div className="usa-alert-body">
         <h2 className="usa-alert-heading">{title}</h2>
-        { children ? <p className="usa-alert-text">{children}</p> :
-          <p className="usa-alert-text">{message}</p>}
+        { children ? <div className="usa-alert-text">{children}</div> :
```
Comment (NickHeiner) [inline client/app/components/Alert.jsx]: Did you check to see if this would cause a problem anywhere else?

File: client/app/intake/pages/finish.jsx [hunk 18]
```diff
@@ -24,8 +24,10 @@ class Finish extends React.PureComponent {
       },
       {
         header: 'VACOLS Issue(s)',
-        valueFunction: (issue) => (
-          issue.description.map((descriptor) => <div>{descriptor}</div>)
+        valueFunction: (issue, index) => (
+          issue.description.map(
+            (descriptor) => <div key={`${descriptor}-${index}`}>{descriptor}</div>
```
Comment (NickHeiner) [inline client/app/intake/pages/finish.jsx]: If we don't have a real `id` to use, we can just use `index` directly.

File: client/app/intake/pages/finish.jsx (hunk 14 above)
Comment (NickHeiner) [reply client/app/intake/pages/finish.jsx]: Ah, I understand what you're saying now. Great.

File: client/app/components/Alert.jsx (hunk 17 above)
Comment (NickHeiner) [reply client/app/components/Alert.jsx]: I see your comment elsewhere that you did.

File: client/app/intake/pages/finish.jsx [hunk 19]
```diff
@@ -36,16 +36,14 @@ class Finish extends React.PureComponent {
       }
     ];
 
-    return _.map(appeals, (appeal) => {
-      return <div key={appeal.id}>
```
Comment (NickHeiner) [inline client/app/intake/pages/finish.jsx]: We still need `key` on `Table` because we are returning an array of elements here:

image

**Comment by NickHeiner** [comment]
I have seen that the feature test fails when the code is broken.

File: client/app/intake/pages/finish.jsx (hunk 19 above)
Comment (shanear) [reply client/app/intake/pages/finish.jsx]: whoops... 🙃 fixed this
